### PR TITLE
[RLlib] Add metrics to IMPALA/APPO/PPO (prototype) to measure off-policy'ness for performed updates.

### PR DESCRIPTION
**Pull request by sven1977**
Signed-off-by: sven1977 <svenmika1977@gmail.com>

Add metrics to IMPALA/APPO (prototype) to measure off-policy'ness for performed updates.
* Adds a `num_grad_updates` counter to the Policy base class. This counter is incremented by 1 by the policy itself upon `learn_on_batch` or `learn_on_loaded_batch`.
* It can be overridden by a global_vars update being pushed down from the Algorithm -> Worker -> Policy.
* Adds `num_grad_updates_per_policy` to global vars of each RolloutWorker.
* The number of grad updates per policy is passed down to sampler policies from the trained policies (after the training step). This happens along with the weights. Note that only those workers that had samples and only those policies that were actually updated, do get this `num_grad_updates` broadcast.
* Uses this new counter to track, how much off-policy'ness we have in our individual updates:
* Off-policy'ness is measured by taking the difference between the number of grad updates that the policy used for sampling has already undergone vs the number of grad updates that the actually trained policy version has already gone through. So for a 100% on-policy algo (like PG), we would expect values of 0.0. For IMPALA and APPO, we should expect small'ish values (between 0.0 and 1.0) and for PPG (with its num_sgd_iter setting), we might even see values >> 1.0, depending on the config, of course.

* Added some unit checks for PPO/IMPALA/APPO unit tests.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/sample_batch.py [hunk 1]
```diff
@@ -103,6 +103,9 @@ def __init__(self, *args, **kwargs):
         self.zero_padded = kwargs.pop("_zero_padded", False)
         # Whether this batch is used for training (vs inference).
         self._is_training = kwargs.pop("_is_training", None)
+        # Average number of grad updates that have been performed
```
Comment (avnishn) [inline rllib/policy/sample_batch.py:106]: is it really the average number of updates? I think its the total sum of gradient updates that have occurred over the life of the trained policy.

We need to update this comment, unless you intended to say "Average" in which case needs to be reflected in the code.

File: rllib/policy/torch_policy.py [hunk 2]
```diff
@@ -467,12 +471,18 @@ def learn_on_batch(self, postprocessed_batch: SampleBatch) -> Dict[str, TensorTy
...
             {
                 "custom_metrics": learn_stats,
                 NUM_AGENT_STEPS_TRAINED: postprocessed_batch.count,
+                NUM_GRAD_UPDATES_LIFETIME: self.num_grad_updates + 1,
+                DIFF_NUM_GRAD_UPDATES_VS_SAMPLER_POLICY: (
+                    self.num_grad_updates - (postprocessed_batch.num_grad_updates or 0)
+                ),
             }
         )
+        self.num_grad_updates += 1
```
Comment (avnishn) [inline rllib/policy/torch_policy.py:485]: move this to before line 75? then compute `DIFF_NUM_GRAD_UPDATES_VS_SAMPLER_POLICY` as `self.num_grad_updates - (postprocessed_batch.num_grad_updates or 0) - 1`

It makes more sense from a readability perspective, because the assumption is that the most recent gradient computation has already happened. 

alternatively, make a new  temporary var called `previous_num_grad_updates` which is `self.num_grad_updates` and use that for computing `DIFF_NUM_GRAD_UPDATES_VS_SAMPLER_POLICY`

**Review by avnishn — COMMENTED** [review]
this change p much makes sense to me. 

I had some minor suggested changes that I think should improve readability of the change

File: rllib/utils/test_utils.py [hunk 3]
```diff
@@ -498,6 +502,54 @@ def check_learning_achieved(
...
+    learner_info = results["info"][LEARNER_INFO]
+    if DEFAULT_POLICY_ID not in learner_info:
+        return None
+    off_policy_ness = learner_info[DEFAULT_POLICY_ID][
+        DIFF_NUM_GRAD_UPDATES_VS_SAMPLER_POLICY
+    ]
+    # Roughly: Reaches up to 0.4 for 2 rollout workers and up to 0.2 for
+    # 1 rollout worker.
+    if not (lower_limit <= off_policy_ness <= upper_limit):
+        raise AssertionError()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/test_utils.py:548]: Yes, you are absolutely right. Forgot to add this. ...

File: rllib/policy/sample_batch.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:106]: Good question. So it's actually the weighted average over the different samples (timesteps) in that batch.

Imagine you have two rollout workers and one of them has a policy that has undergone 2 updates thus far, the other uses a policy that has undergone 3 updates thus far. Both workers produce rollouts of length 50 (rollout_fragment_length==50). The train batch size is 100, so we concatenate these 2x len=50 batches to one that's 100 ts long. This new 100ts batch will have its `num_gradient_updates` property set to 2.5 as it's the weighted average (both original batches contribute 50%).

Let me try making this more clear in the comment here ..

File: rllib/utils/test_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/test_utils.py:548]: done

File: rllib/policy/sample_batch.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:106]: Enhanced the comment to make this more clear.

File: rllib/policy/torch_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/torch_policy.py:485]: You are right, we should increase the counter as soon as we do the actual update. Fixed.

File: rllib/policy/sample_batch.py (hunk 1 above)
Comment (avnishn) [reply rllib/policy/sample_batch.py]: oh I see. sg

**Review by avnishn — APPROVED** [review]
